File: DONGUK/프로그래머스_WEEK 4/조이스틱.py
def solution(name):
    answer = 0
    
    # 기본 최소 좌우이동 횟수는 길이 - 1
    min_move = len(name) - 1
    
    for i, char in enumerate(name):
        answer += min(ord(char) - ord('A'), ord('Z') - ord(char) + 1)
        
        # 해당 알파벳 다음부터 연속된 A 문자열 찾기
        next = i + 1
        while next < len(name) and name[next] == 'A':
            next += 1
            
        # 기존, 연속된 A의 왼쪽시작 방식, 연속된 A의 오른쪽시작 방식 비교 및 갱신
        min_move = min([min_move, 2 *i + len(name) - next, i + 2 * (len(name) -next)])
        
    # 알파벳 변경(상하이동) 횟수에 좌우이동 횟수 추가
    answer += min_move
    return answer

# 71점 받는 풀이 -> 현재 위치에서 변경해야 할 가장 가까운 인덱스를 찾아서 이동하는 풀이인데 안됨
# def solution(name):
#     answer = 0
#     N = len(name)
    
#     ind_list = []
    
#     target = 'A' * N
    
#     for i in range(N):
#         if name[i] != target[i]:
#             diff_a = abs(ord(name[i]) - ord(target[i]))
#             diff_b = abs(ord('Z') - ord(name[i]) + 1)
#             ind_list.append(min(diff_a, diff_b))
#         else:
#             ind_list.append(0)
    
    
#     if not ind_list:
#         return answer
    
    
#     answer += sum(ind_list)
    
#     ind = 0
#     while True:
#         ind_list[ind] = 0
        
#         if sum(ind_list) == 0:
#             break

#         left, right = 1, 1
#         while ind_list[ind - left] == 0:
#             left += 1
        
#         while ind_list[ind + right] == 0:
#             right += 1
        
#         if left < right:
#             answer += left
#             ind -= left
            
#         else:
#             answer += right
#             ind += right
        
#     return answer

File: JINSEOP/프로그래머스_WEEK4/조이스틱.cpp
#include <string>
using namespace std;

int up_down(char part_name) {
	if (part_name - 'A' < 14)
		return part_name - 'A';
	else return 26 - ((part_name) -'A');
}

int solution(string name) {
	int answer = 0;
    int turn = name.size()-1; // 커서가 한 쪽 방향으로만 가는 경우
    
    for(int i = 0; i<name.size(); i++){
		if(name[i] == 'A') continue; // A인 자리는 검사할 필요 없음
        answer += up_down(name[i]);
        int index = i+1; // 다음 바꿀 index
        while(index < name.size() && name[index] == 'A')
            index++; // A가 아닌 다음 문자 찾기
        
        int a = i; // 원점 ~ 현재 index
        int b = name.size() - index; // 배열 끝 ~ 다음 바꿀 index
        turn = min(turn, min(2*a+b, a+2*b));
		// (a + a) + b : 시작점 -> i -> 시작점 -> index
		// (b + b) + a : 시작점 -> index -> 시작점 -> i
    }
    
    if(answer != 0) answer += turn; // A로만 구성된 name일 때,
				    // turn을 더해주지 않도록.
	return answer;
}
